File: framework/languages/java/e2e-kotlin/SpotService/Client/src/main/kotlin/systems/zlink/e2e/kotlin/spotservice/client/scenarios/SmB3Scenario.kt
package systems.zlink.e2e.kotlin.spotservice.client.scenarios

import systems.zlink.e2e.kotlin.spotservice.client.support.ActorSessionScenarioContext
import systems.zlink.e2e.kotlin.spotservice.client.support.ensure

internal object SmB3Scenario {
    suspend fun run(context: ActorSessionScenarioContext) {
        ensure(context.auth.displayName == context.profile.displayName, "SM-B3 create profile display name mismatch")
        ensure(context.auth.level == context.profile.level, "SM-B3 create profile level mismatch")
        ensure(context.auth.tags == context.profile.tags, "SM-B3 create profile tags mismatch")

        ensure(context.entryReply.requestSeq == 1, "SM-B3 entry request sequence mismatch")
        ensure(context.entryReply.displayName == context.profile.displayName, "SM-B3 entry profile display name mismatch")
        ensure(context.entryReply.level == context.profile.level, "SM-B3 entry profile level mismatch")
        ensure(context.entryReply.tags == context.profile.tags, "SM-B3 entry profile tags mismatch")

        ensure(context.joined.tags == context.profile.tags, "SM-B3 join payload tags mismatch")
        ensure(context.joined.displayName == context.profile.displayName, "SM-B3 join payload display name mismatch")
        ensure(context.joined.level == context.profile.level, "SM-B3 join payload level mismatch")

        ensure(context.userReply1.requestSeq == 2, "SM-B3 user request sequence mismatch")
        ensure(context.userReply1.displayName == context.profile.displayName, "SM-B3 user profile display name mismatch")
        ensure(context.userReply1.level == context.profile.level, "SM-B3 user profile level mismatch")
        ensure(context.userReply1.tags == context.profile.tags, "SM-B3 user profile tags mismatch")

        println("scenario SM-B3 passed")
    }
}

File: framework/languages/node/e2e/SpotService/Client/Scenarios/sm-b3-scenario.ts
// SM-B3: Typed Actor request payload를 보존한다 시나리오를 검증한다.
import {
  zlinkStreamConnectorFactory,
  zlinkStreamJsonCodec,
  ZlinkStreamDispatchMode
} from '@zlink-systems/stream-connector';
import type {
  AuthRes,
  AuthReq,
  ComplexActorRes,
  ComplexActorReq,
  EvidenceWaitReq
} from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { postJson } from '../../../http-client';
import { ensure } from '../Support/scenario-assert';

export async function runSmB3(options: ClientOptions): Promise<void> {
  const actorId = `actor-sm-b3-complex-${Date.now()}`;
  const client = zlinkStreamConnectorFactory.create({
    endpoint: options.sessionAStreamEndpoint,
    codec: zlinkStreamJsonCodec,
    dispatchMode: ZlinkStreamDispatchMode.Immediate,
    heartbeat: { enabled: false },
    waitTimeoutMs: 10000
  });
  await client.connect();
  try {
    await client
      .request({
        actorId,
        displayName: 'complex actor',
        nodeRid: 'play-a'
      } satisfies AuthReq)
      .packetName('AuthReq')
      .timeout(5000)
      .submit<AuthRes>();

    const complex = await client
      .request({
        displayName: 'Ada Lovelace',
        level: 42,
        tags: ['alpha', 'beta', 'gamma'],
        attributes: {
          role: 'analyst',
          region: 'west'
        }
      } satisfies ComplexActorReq)
      .packetName('ComplexActorReq')
      .timeout(5000)
      .submit<ComplexActorRes>();

    ensure(complex.actorId === actorId, 'SM-B3 actor id mismatch.');
    ensure(
      complex.displayName === 'Ada Lovelace' && complex.level === 42,
      'SM-B3 scalar payload mismatch.'
    );
    ensure(complex.tags.join(',') === 'alpha,beta,gamma', 'SM-B3 tag payload mismatch.');
    ensure(
      complex.attributes.role === 'analyst' && complex.attributes.region === 'west',
      'SM-B3 attribute payload mismatch.'
    );

    const expectedEvidence = [`actor-complex|rid=play-a|actor=${actorId}`];
    const evidence = await postJson<string[]>(options.playAUrl, '/evidence/wait', {
      containsAll: expectedEvidence,
      timeoutMilliseconds: 10000
    } satisfies EvidenceWaitReq);
    ensure(
      expectedEvidence.every((expected) => evidence.some((line) => line.includes(expected))),
      'SM-B3 evidence mismatch.'
    );
  } finally {
    await client.close();
  }

  console.log('scenario SM-B3 passed');
}
